client/client.go

```diff
@@ -2561,3 +2597,21 @@ func (c *Client) allAllocs() map[string]*structs.Allocation {
 	}
 	return allocs
 }
+
+// group wraps a func() in a goroutine and provides a way to block until it
+// exits. Inspired by https://godoc.org/golang.org/x/sync/errgroup
+type group struct {
+	wg sync.WaitGroup
+}
+
+func (g *group) Go(f func()) {
+	g.wg.Add(1)
+	go func() {
+		defer g.wg.Done()
+		f()
+	}()
+}
+
+func (g *group) Wait() {
+	g.wg.Wait()
```

Comment:
You caught me. I should have called this out in a comment. It is exactly why I made this private as it should be "Good Enough" for the Client's needs but not robust for general use. We're faced with 3 choices:

1. Consider this Good Enough for the Client and move on
2. Remove this and DIY blocking on the small number of goroutines we actually need to block on (many can just be canceled and assumed to stop at some point in the future)
3a. Make this more robust and suitable for general use
3b. Use a 3rd party library with the characteristics we need (tomb?)

I'm taking the Thanksgiving break to mull it over but feedback/opinions appreciated!

Reply:
Yeah, good catch. As I mentioned above I'm going to remove uses of the group outside of NewClient to avoid the race. Luckily the alloc syncing related goroutines are the ones we care about the most and they're started from NewClient, so they're safe to keep in the group.